RFC: Introduce a STM32 SYSCFG interrupt controller driver

eraquin-kalray opened the pull request:
This PR is a follow up of [RFC: Introduce a STM32 SYSCFG interrupt controller driver](https://github.com/zephyrproject-rtos/zephyr/pull/57310).

The advantage of this solution over current shared interrupts implementation provided by https://github.com/zephyrproject-rtos/zephyr/pull/61422 is that only relevant ISR is called.
Whereas the current solution is calling each and every ISR sharing the IT, regardless of whether there is a pending IT for the device or not. This can results in error if ISR of device driver relying on this solution isn't safe to be called without any pending IT. For instance `drivers/pwm/pwm_stm32.c::pwm_stm32_isr()` has some NULL pointer that aren't checked in this case.

Commits details :

- First 3 commits are a split of the first one of the original RFC from @aurel32, as requested during review.
- Purpose of commits 4 and 5 is to fix the issue highlithed by @aurel32 in [this comment](https://github.com/zephyrproject-rtos/zephyr/pull/57310#issuecomment-1529158470).
  Based on existing implementation for other socs, like the RV32M1, relying on SOC specific IT management (`ARM_CUSTOM_INTERRUPT_CONTROLLER`), which is selected when enabling interrupt controller driver on STM32
- Commits 6 and 7 are there to fix incorrect handling of multi level interrupts number when getting from name.
  Identified when validating the implementation using CAN counter sample code. `DT_INST_IRQ_BY_NAME` wasn't returning the expected multi level encoded interrupt number. Proposal is to add devicetree helpers to get (multi level) interrupt number from name, like existing ones from index.
- Commit 8 is adding a sample code to show how to setup interrupts to use SYSCFG IT lines.
  As discussed during review, IT line muxes nodes aren't enabled in `.dtsi`. Thus interrupts settings aren't updated to use SYSCFG IT lines, and no configuration is performed in this way. This sample code shows what has to be done on board `.dts` and project configuration sides.

Implementation tested by building and checking generated `isr_tables.c::_sw_isr_table[]`, with and without SYSCFG and IT line muxes. Tested at runtime as well on NUCLEO-G0B1RE using CAN counter sample code.

Any suggestion welcomed. Please tell me if something should be added, removed, implemented another way !

eraquin-kalray committed:
soc: st: stm32: Add SOC specific IRQ manager to support interrupt muxer

Add STM32 SOC speficic IRQ manager to support usage of interrupt muxer
when enabling and disabling interrupts

Signed-off-by: Etienne Raquin <eraquin@kalrayinc.com>

eraquin-kalray committed:
drivers: interrupt_controller: Selects ARM custom IT CTRL on STM32

Update interrupt controller driver config for STM32, so that it selects
ARM custom interrupt controller, which is required to handle interrupts
enable/disable

Signed-off-by: Etienne Raquin <eraquin@kalrayinc.com>

eraquin-kalray committed:
devicetree: Add missing helpers to get multi level IT number by name

Add missing devicetree helpers to get interrupt number from name when
using multi level interrupts

Signed-off-by: Etienne Raquin <eraquin@kalrayinc.com>

eraquin-kalray committed:
drivers: can: stm32: fdcan: Fix IRQ connect by name on multi level

Fix STM32 canfd driver interrupt connect by name when using multi level
interrupts, by using new DT_INST_IRQN_BY_NAME helper macro

Signed-off-by: Etienne Raquin <eraquin@kalrayinc.com>

eraquin-kalray commented:
As I feel like failing test isn't related to my changes, marking as ready for review

eraquin-kalray commented on dts/bindings/interrupt-controller/st,stm32-syscfg-itline.yaml:
From my current understanding, not having its own compatible would result in a `intc_syscfg_stm32.c` driver at SYSCFG level, and a single `st_stm32_syscfg` compatible instance, similar to existing `intc_rv32m1_intmux.c`/`openisa_rv32m1_intmux`.
Is it what you are suggesting to go for ?

eraquin-kalray replied:
Thanks for confirmation, I will have a look

eraquin-kalray replied:
"Interrupt Line" are now child nodes of `st,stm32-syscfg`, without any `compatible`
`intc_syscfg_itline_stm32.c` driver with multiple instances turned into `intc_syscfg_stm32.c` with a single instance, managing all children IT lines
Which allows to get rid of ugly way of getting reference to IT line device at runtime

eraquin-kalray commented on dts/arm/st/g0/stm32g0b0.dtsi:
```diff
@@ -119,3 +121,21 @@
+};
+
+&itline26_mux {
+	status = "okay";
```
Ok fine
If not activating IT line in `.dtsi`, does it make sense to keep IT line being set as IT parent, like following one, in the `.dtsi` ?

```dts
&i2c2 {
	interrupt-parent = < &itline24_mux >;
	interrupts = <0 0>;
};
```

Asking because it wouldn't work without IT line being activated in the board `.dts`
Do you suggest to move all this IT related settings in the board `.dts` ?
Or is it better to make an exception, and keep them in the `.dtsi`, like in `dts/riscv/openisa/rv32m1_ri5cy.dtsi` ?

eraquin-kalray replied:
It would, if child-nodes aren't featuring any `status = "disabled"`
Which results in having them enabled anyway, even if not used`
Thus, `_sw_isr_table` is populated with syscfg ISR for each and every IT line
Setting up a device IT without using syscfg itline results in a shared IT between syscfg and device driver ISRs
(Hope it's clear enough)
I don't think it's the expected behavior
Maybe better not to change any IT setting in `dtsi`, and let the user update as-needed in board `.dts` ?

> make the IT lines child-nodes

Implemented locally, syscfg driver updated accordingly, working as expected
Still need to address [your 3rd comment](https://github.com/zephyrproject-rtos/zephyr/pull/82776#discussion_r1879589884) before pushing

eraquin-kalray replied:
Updated accordingly
No enabling of IT line mux nodes in `.dtsi`. Removed IT 2nd level related settings in configuration files, as not applicable anymore
Created a sample code to show how to setup board `.dts` and project configuration to use SYSCFG IT lines

erwango commented:
Initial question on my side :

What is the advantage of this new solution above https://github.com/zephyrproject-rtos/zephyr/pull/61422 which is currently in use to handle IRQ sharing on STM32G0 for instance ? And why we would want to support both ?

PR description should detail this.

eraquin-kalray commented:
> Initial question on my side [...]

Forgot to highlight it, my bad

> PR description should detail this

PR description updated accordingly

eraquin-kalray committed:
dts: bindings: stm32: Add STM32 SYSCFG INTC driver DTS binding

Add DTS binding for STM32 SYSCFG interrupt controller driver

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>
Signed-off-by: Etienne Raquin <eraquin@kalrayinc.com>

eraquin-kalray committed:
drivers: interrupt_controller: introduce a STM32 SYSCFG INTC driver

This commit tries to solve the problem of the shared interrupt on the
STM32G0 line of MCUs (and probably other lines) by adding an interrupt
muxer for lines that are shared by multiple devices, leveraging the
existing multilevel interrupts framework.

It uses flag pending interrupts from each interrupt line available in
the SYSCFG controller to correctly route the interrupts to the correct
device, and correctly enable or disable the shared interrupt depending
on the state of each individual interrupt.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>
Signed-off-by: Etienne Raquin <eraquin@kalrayinc.com>

eraquin-kalray committed:
dts: arm: st: g0: Add stm32-syscfg node

Add STM32 SYSCFG node to ARM STM32G0 DTS

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>
Signed-off-by: Etienne Raquin <eraquin@kalrayinc.com>

eraquin-kalray committed:
samples: boards: stm32: Add STM32 SYSCFG IT line sample

Add sample code for NUCLEO-G0B1RE to show how to setup interrupts using
STM32 SYSCFG IT line

Signed-off-by: Etienne Raquin <eraquin@kalrayinc.com>

eraquin-kalray commented on dts/bindings/misc/st,stm32-syscfg.yaml:
Now splitted in 3 commits

eraquin-kalray commented:
Sorry for repeated force pushes, struggling a little bit with compliance checks
Now passing locally, will re-request review once online check reporting OK

eraquin-kalray commented:
@JarmouniA Did you get a chance to have another look after my update ?

eraquin-kalray commented:
@KozhinovAlexander I had a quick look at your PR, I don't see any collision either
As you didn't update the way `IRQ_CONNECT` is performed, having SYSCFG itline as interrupt parent of the EXTI should work as-is, if needed

mathieuchopstm commented on samples/boards/st/stm32_syscfg_itline/Kconfig:
```diff
@@ -0,0 +1,38 @@
+config 2ND_LVL_ISR_TBL_OFFSET
+	default 32
+
+config 2ND_LVL_INTR_00_OFFSET
+	default 21
+
```
I stand corrected: `gen_isr_tables.py` uses them here:
https://github.com/zephyrproject-rtos/zephyr/blob/459a078d22607a0007ff3f4a2ef31c29c71a162f/scripts/build/gen_isr_tables.py#L112-L128

This also means we can't have more than 8 muxed devices at a time, as far as I understand this... maybe `gen_isr_tables` could be improved to automagically figure this out (or just not need it), but that's outside this PR's scope.

eraquin-kalray commented on drivers/interrupt_controller/Kconfig.stm32:
```diff
@@ -19,4 +19,14 @@ config GPIO_INTC_STM32WB0
 	help
 	  Enable GPIO interrupt controller driver for STM32WB0 series
 
+config SYSCFG_STM32
+	bool "System Configuration Controller (SYSCFG) Driver for STM32"
```
Agreed. Got same remark on our side
Will update both config name and description

eraquin-kalray commented on drivers/interrupt_controller/Kconfig.stm32:
```diff
@@ -19,4 +19,14 @@ config GPIO_INTC_STM32WB0
+	depends on DT_HAS_ST_STM32_SYSCFG_ENABLED
+	select MULTI_LEVEL_INTERRUPTS
+	select 2ND_LEVEL_INTERRUPTS
+	select ARM_CUSTOM_INTERRUPT_CONTROLLER
```
> `MULTI_LEVEL_INTERRUPTS` already enables SoC-specific hooks

Right. Missed it as it isn't the case at revision we are using on our fork
SoC-specific IRQ management source file should be way lighter this way, updating

eraquin-kalray commented on drivers/interrupt_controller/intc_syscfg_stm32.c:
> I think this is mostly due to the `st_stm32_syscfg device` [...] becoming the manager of the individual ITMUX devices

Agreed. Which requires `itline_devs` and `syscfg_register_itline()` to work

> Is there a reason for this architecture [...] ?

Main reason is rework suggested/requested in [this comment](https://github.com/zephyrproject-rtos/zephyr/pull/82776#discussion_r1879559483)

I do agree that `intc_cavs.c` and `intc_dw` and look simpler
I also prefer "core" interrupts being enabled/disabled in the SoC-specific interrupt manager instead of the driver

Before rework, implementation was closer to #57310 one
Should we go back to it ? Having driver instances at SYSCFG itline level
I feel like decision has to be made depending on what the DTSI should look like

mathieuchopstm replied:
I agree that a single binding with `child-properties` would be cleaner.

I don't see anything preventing instantiation of only the children (i.e., keep `DT_INST_FOREACH_CHILD_STATUS_OKAY(0, ...)` and drop `DEVICE_DT_INST_DEFINE(0, ...)`) - couldn't we go in this direction?
```dts
syscfg {  /* this node is not instantiated */
  compatible = "st,g0-syscfg";

  /* these nodes are instantiated by SYSCFG driver as 2nd-level IRQ controllers */
  itlinemux0 { };
  itlinemux1 { };
  /* ... */
};
```

I'm not familiar yet with how the whole multi-level IRQs work in Zephyr so maybe I'm missing something that makes this not possible... However, `intc_cavs.c` is used in almost the same way as what I'm proposing - the only difference is that it uses compatible-based instancing, rather than child-based - so I'm hoping it should be fine?

Wasn't aware of #57310 - will take a look.

eraquin-kalray commented on dts/arm/st/g0/stm32g0.dtsi:
```diff
@@ -447,6 +447,237 @@
+			reg = <0x40010000 0x400>;
+			clocks = <&rcc STM32_CLOCK_BUS_APB1_2 0x1>;
+
+			itline0_mux: itline0_mux {
+				#interrupt-cells = <2>;
+				interrupt-controller;
+				interrupts = <0 0>;
+				status = "disabled";
+			};
```
Agreed for IT lines that aren't shared
I do agree for EXTI as well, from what I got when having a look at #85508
Need to have a closer look for DMA ones

Regarding lines 2 and 4, didn't get your point
What do you mean by "is managed by" ?

mathieuchopstm replied:
What I meant by "is managed by" is that (as far as I remember) the RTC driver is responsible for handling the `TAMP` interrupt, and the RCC driver is responsible for handling the `CRS` interrupt, so they wouldn't need someone else to do the muxing on their behalf.

henrikbrixandersen commented on drivers/can/can_stm32_fdcan.c:
Please avoid commits combining white-space and functional changes like this. It's very difficult to read.

henrikbrixandersen commented:
Please rebase on `main`, resolve any merge conflicts, and address the review comments.
